src/mvba/vcbc/mod.rs

```diff
@@ -1,75 +1,325 @@
+                                    tag: self.tag.clone(),
+                                    action: MSG_ACTION_C_READY.to_string(),
+                                    m: self.d.as_ref().unwrap().to_bytes(),
+                                    sig: Sig::SignatureShare(s1),
+                                };
 
-        self.state.process_message(sender, &msg, &mut self.ctx)?;
-        if let Some(s) = self.state.decide(&mut self.ctx)? {
-            self.state = s;
+                                // send (ID.j.s, c-ready, H(m), ν) to Pj
+                                self.send_to(&ready_msg, self.tag.j)?;
+
+                                self.state = State::Ready;
+                                return self.decide();
+                            }
+                        }
+                        Ok(())
+                    }
+                    None => Ok(()),
+                }
+            }
+            State::Ready => {
+                match self.message_log.get(MSG_ACTION_C_READY) {
+                    Some(msgs) => {
+                        // TODO: better way?
+                        let msg_cloned = msgs.clone();
+                        for (l, msg) in msg_cloned {
+                            if !msg.m.eq(&self.d.as_ref().unwrap().as_fixed_bytes()) {
+                                warn!(
+                                    "c-ready has unknown digest. expected {:?}, got {:?}",
+                                    self.d.as_ref().unwrap(),
+                                    msg.m
+                                );
+                                continue;
+                            }
+                            let sign_bytes = self.sign_bytes();
+
+                            // Upon receiving message (ID.j.s, c-ready, d, νl) from Pl for the first time:
+                            if let Vacant(e) = self.wd.entry(l) {
+                                if let Sig::SignatureShare(sig_share) = msg.sig {
+                                    let valid_sig = self
+                                        .pub_key_set
+                                        .public_key_share(l)
+                                        .verify(&sig_share, sign_bytes);
+
+                                    if !valid_sig {
+                                        warn!("c-ready has has invalid signature share");
+                                    }
+
+                                    // if i = j and νl is a valid S1-signature share then
+                                    if self.i == msg.tag.j && valid_sig {
+                                        // Wd ← Wd ∪ {νl}
+                                        e.insert(sig_share);
+
+                                        //  rd ← rd + 1
+                                        self.rd += 1;
+
+                                        // if rd = n+t+1/2 then
+                                        if self.rd == (self.number + self.threshold + 1) / 2 {
```

Comment:
The threshold value should be coming from the pub-key-set:
```suggestion
                                        if self.rd > self.pub_key_set.threshold() {
```